specs/altair/p2p-interface.md

```diff
@@ -94,6 +112,92 @@ Unaggregated signatures (along with metadata) are sent as `SyncCommitteeSignatur
 
 Aggregated sync committee signatures are packaged into (signed) `SyncCommitteeContribution` along with proofs and gossiped to the `sync_committee_contribution_and_proof` topic.
 
+### Transitioning the gossip
+
+With any fork, the fork version, and thus the `ForkDigestValue`, change.
+Message types are unique per topic, and so for a smooth transition a node will have to subscribe to both the old and new topics.
+
+All remaining topics are updated. In advance of the fork, a node SHOULD subscribe to the post-fork variants of the topics.
+
+Subscriptions are expected to be well-received, all updated nodes should subscribe as well.
+Topic-meshes can be grafted quickly as the nodes are already connected and exchanging gossip control messages.
+
+Messages SHOULD NOT be re-broadcast from one fork to the other.
+
+Pre-fork:
+- Peers who propagate messages on the post-fork topics MAY be scored negatively proportionally to time till fork,
+  to account for clock discrepancy.
+- Messages can be IGNORED on the post-fork topics, with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` margin.
+
+Post-fork:
+- Peers who propagate messages on the pre-fork topics MUST NOT be scored negatively. Lagging IWANT may force them to.
+- Messages on pre and post-fork variants of topics share application-level caches.
+  E.g. an attestation on the both the old and new topic is ignored like any duplicate.
+- Two epochs after the fork, pre-fork topics SHOULD be unsubscribed from. This is well after the configured `seen_ttl`.
+
+## The Req/Resp domain
+
+### Req-Resp interaction
+
+An additional `<context-bytes>` field is introduced to the `response_chunk` as defined in the phase0 document:
+
+```
+response_chunk  ::= <result> | <context-bytes> | <encoding-dependent-header> | <encoded-payload>
+```
+
+All phase0 methods are compatible: `<context-bytes>` is empty by default.
+On a non-zero `<result>` with `ErrorMessage` payload, the `<context-bytes>` is also empty.
+
+In Altair and later forks, `<context-bytes>` functions as a short meta-data,
+defined per req-resp method, and can parametrize the payload decoder.
+
+#### `ForkDigest`-context
+
+Starting with Altair, and in future forks, SSZ type definitions may change.
+For this common case, we define the `ForkDigest`-context:
+
+A fixed-width 4 byte `<context-bytes>`, set to the `ForkDigest` matching the chunk:
+ `compute_fork_digest(fork_version, genesis_validators_root)`.
+
+### Messages
+
+#### BeaconBlocksByRange v2
```

Comment:
How does adding the `context` metadata work here ? They are only included in response packets rather than requests. Is the goal for consumers to be aware of the particular fork version of the response payload ?

Reply:
The version bump is to distinguish between an empty `context`, and a `ForkDigest` used in v2. The format is "compatible" in the sense that we can define the format of the old v1 methods as a subset of the new format, by simply assuming an empty context. But the actual change of a method is breaking, hence a v2.

The idea is that everyone can update their Req-Resp stack with this small change, and keep supporting the old formats almost effortlessly. Most methods don't need a v2 with Altair. Just in the case of these two beacon block methods, the fork digest can now be read before touching the SSZ decoder, which should make it easy to handle different block types.

Regarding implementation, the verbose way of describing it would be:
1. Add a `CodecContext` interface/base-type/whatever preferred.
2. Parametrize the response writer with `CodecContext`
3. Specify `EmptyContext` as a subtype/implementation/whatever of `CodecContext`, with a read/write method, that is a no-op.
4. Specify that all v1 request handlers have an `EmptyContext` as context type
5. Specify that all v1 response handlers have an `EmptyContext` as context type
6. Specify `ForkDigestContext` as subtype/implementation/whatever of `CodecContext`, with a read/write method, that reads/writes 4 bytes into/from a fork digest.
6. Define the new v2 block request handlers in terms of the old v1 handlers, but parametrized with `ForkDigestContext`. The chunk writer can switch on the type of context, and when it's `ForkDigestContext`, it writes the fork digest corresponding to the block.
7. Define the new v2 block response handlers in terms of the old v1 handlers, but parametrized with `ForkDigestContext`. The chunk reader first reads the expected 4 byte fork digest, validates if it's a known fork, and then continues to decode the chunk into the expected type of block.

We can also rename the req-resp `CodecContext` to `CodecParameter` or something if it's confusing in Go.